### DISCUSSION TITLE
SPS w/ AP2 build in docker?

### DISCUSSION THREAD

**Discussion by stephenmetzger** [mikebrady/shairport-sync#1538] [Q&A]
Hello, 

I have multiple instances of SPS running and would like to reinstall using docker instead of the 'normal' installation and AP2 support. 

I build my instances with the --with-mqtt-client and --with-metadata options. I need these, but am not sure how to implement in a docker or docker-compose build. Is there any other documentation regarding advanced configuration of SPS with AP2 and *--with options?

**Comment by charlesomer** [comment]
Hi, have you seen the following README? https://github.com/mikebrady/shairport-sync/blob/development/docker/README.md

There's a prebuilt docker image available with those options based on the development branch 😄 

However, I'm not sure about the current state of multiple instances on one device (AirPlay 2) and if that works or not

**Reply by mikebrady** [reply]
Unfortunately, multiple instances of the AirPlay2 version of Shairport Sync on one device will not work -- this is an apparent limitation of the AirPlay 2 players.

But please see Issue #1452 for more.

**Reply by stephenmetzger** [reply]
Thanks @charlesomer and @mikebrady. I have looked at the README. I see 

```To get access to the full range of configuration options, pass the configuration file to /etc/shairport-sync.conf in the container using the -v option or docker compose.```

But, in using a normal build, I typically build it with the *--with options, and then have to go into the .conf to configure things like my MQTT broker settings, and other options. The .conf file states


```For this section to be operative, Shairport Sync must be built with at one (or more) of the following configuration flags: --with-metadata, --with-dbus-interface, --with-mpris-interface or --with-mqtt-client.```

Can you point me to the pre-built docker image with those build options? I cannot seem to find it.

**Reply by mikebrady** [reply]
Sure. The images at https://hub.docker.com/repository/docker/mikebrady/shairport-sync are built with those build options.

Please note that there is a bug in the AirPlay 2 images -- command line arguments added to the end of the command to run the image are ignored at present.

**Reply by stephenmetzger** [reply]
Thanks @mikebrady. 

I have the install built, and pointed to mounted volume with my shairport-sync.conf saved. 

On container start, I'm getting:

```*warning: could not acquire a Shairport Sync native D-Bus interface "org.gnome.ShairportSync.i142" on the system bus.```

and 

```warning: could not acquire an MPRIS interface named "org.mpris.MediaPlayer2.ShairportSync.i142" on the system bus.```

The container never seems to truly start (can't exec in), and does not show up as an airplay source. 

I did have a standard install (non-docker) on this RPi3B+, using the headphone jack with no DAC or anything, prior to this docker install. I used the which-shairport-sync command to find and delete copies and deleted service files. 

Any thoughts on what might be causing these errors?

**Reply by charlesomer** [reply]
Hi, could you try pulling the `latest` image please? (`mikebrady/shairport-sync:latest`). I'm guessing you're only trying to run the one instance?

**Reply by stephenmetzger** [reply]
Thanks @charlesomer. I was using the `unstable-development` image.  Does the `latest` image support AirPlay 2?

Yes, one instance on each of the RPi's I have running SPS.

**Reply by charlesomer** [reply]
It does now yes 😊 (this was changed recently)

**Reply by stephenmetzger** [reply]
Fantastic re: AP2 being pushed to the stable build. 

Still getting the `D-Bus` and `MPRIS` errors on container startup, however. 

Edit:
If it helps at all, this is my docker-compose.yml. Didn't change much. Does the devices section need to be configured different for my use-case (RPi 3B+ with no DAC, running audio straight out of the 3.5mm jack)?

```services:
  shairport-sync:
    image: mikebrady/shairport-sync:latest
    network_mode: host
    restart: unless-stopped
    devices:
      - "/dev/snd"
    volumes:
      - /data/shairport-sync/shairport-sync.conf:/etc/shairport-sync.conf
    logging:
      options:
        max-size: "200k"
        max-file: "10"

**Reply by charlesomer** [reply]
That looks okay to me...

It sounds like you've followed the instructions [here](https://github.com/mikebrady/shairport-sync/blob/development/BUILD.md#remove-old-copies-of-shairport-sync) to remove service files etc. of your old install and rebooted?
